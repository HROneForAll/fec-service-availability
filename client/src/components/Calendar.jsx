import React from 'react';
import Week from './Week.jsx';
import convertToMonth from '../scripts/monthConverter.js';

class Calendar extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      start: null,
      end: null,
      weeks: [], 
      month: null,
      year: null
    };
  }

  initialStart(callback) {
    const today = new Date;
    let day = today.getDay();
    let date = today.getDate();
    while ((date - 1) % 7 !== 0 || date !== 1) {
      date--;
      day--;
      day < 0 ? day = 6 : day;
    }
    this.setState({ start: day });
    callback ? callback(day) : null;
  }

  findFirstOfMonth(weeks) {
    for (let i = 0; i < 7; i++) {
      if (weeks[0][i] !== null) {
        return i;
      }
    }
    return 6;
  }

  findEndOfMonth(weeks) {
    const len = weeks.length - 1;
    for (let i = 0; i < 7; i++) {
      if (weeks[len][i] === null) {
        return i - 1;
      }
    }
    return 6;
  }

  setYearAndMonth() {
    const today = new Date;
    const year = today.getFullYear();
    const month = today.getMonth();
    this.setState({year, month});
  }

  buildWeeks(start, end, year, month) {
    const now = new Date; 
    year = year ? year : now.getFullYear();
    month = month ? month : now.getMonth();
    const days = Object.keys(this.props.dates[year][month]);
    if (start !== null) {
      this.buildFromStart(days, start);
    } else if (end !== null) {
      this.buildFromEnd(days, end);
    } else {
      console.log('Something went wrong...');
    }
  }

  buildFromStart(days, start) {
    let weeks = [];
    let week = [];
    let count = 0;
    while (count < days.length + start) {
      if (count < start) {
        week.push(null);
      } else {
        week.push(days[count - start]);
      }
      if (week.length === 7) {
        weeks.push(week);
        week = [];
      }
      if (count === days.length + start - 1 && week.length < 7 && week.length) {
        while (week.length < 7) {
          week.push(null);
        }
        weeks.push(week);
      }
      count++;
    }
    const end = this.findEndOfMonth(weeks);
    this.setState({ weeks, end, start });
  }

  buildFromEnd(days, end) {
    let weeks = [];
    let week = [];
    let count = days.length + 6 - end;
    while (count > 0) {
      if (count > days.length) {
        week.unshift(null);
      } else {
        week.unshift(days[count - 1]);
      }
      if (week.length === 7) {
        weeks.unshift(week);
        week = [];
      }
      if (count === 1 && week.length && week.length < 7) {
        while (week.length < 7) {
          week.unshift(null);
        }
        weeks.unshift(week);
      }
      count--;
    }
    const start = this.findFirstOfMonth(weeks);
    this.setState({ weeks, start, end });
  }

  plusMonth() {
    let month = this.state.month;
    let year = this.state.year;
    month++;
    if (month > 11) {
      month = 0;
      year++;
      this.setState({ month, year });
    } else {
      this.setState({ month });
    }
    let end = this.state.end;
    end++;
    end = end > 6 ? 0 : end;
    this.buildWeeks(end, null, year, month);
  }

  minusMonth() {
    let month = this.state.month;
    let year = this.state.year;
    month--;
    if (month < 0) {
      month = 11;
      year--;
      this.setState({ month, year });
    } else {
      this.setState({ month });
    }
    let start = this.state.start;
    start--;
    start = start < 0 ? 6 : start;
    this.buildWeeks(null, start, year, month);
  }

  render() {
    const month = convertToMonth(this.state.month);
    const year = this.state.year;
    return (
      <div className="calendar">
        <div>
          <div className="previous-month-button" onClick={() => this.minusMonth()}><i className="fas fa-arrow-left"></i></div>
          <div className="month-header">{month} {year}</div>
          <div className="next-month-button" onClick={() => this.plusMonth()}><i className="fas fa-arrow-right"></i></div>
        </div>
        <div>
          <table>
            <tbody>
              <tr>
                <th>Su</th>
                <th>Mo</th>
                <th>Tu</th>
                <th>We</th>
                <th>Th</th>
                <th>Fr</th>
                <th>Sa</th>
              </tr>
              {this.state.weeks.length
                ? this.state.weeks.map((week, ind) => <Week key={ind} week={week}/>)
                : null}
            </tbody>
          </table>
        </div>
      </div>
    );
  }

  componentDidMount() {
    this.initialStart((start) => this.buildWeeks(start));
    this.setYearAndMonth();
  }

}

export default Calendar;