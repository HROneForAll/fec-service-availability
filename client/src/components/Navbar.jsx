import React from 'react';
import Summary from './Summary.jsx';
import Price from './Price.jsx';

class Navbar extends React.Component {
  constructor(props) {
    super(props);
  }

  render() {
    const date = new Date;
    // console.log(this.props.year);
    const price = this.props.year[date.getMonth()][date.getDate()]['price'];
    // console.log(price);

    return (
      <div>
        <Summary type={this.props.type} location={this.props.location}
        stars={this.props.stars} reviews={this.props.reviews}/>
        <Price price={price}/>
        <Book click={this.props.click}/>
      </div>
    );
  }
    
}

export default Navbar;