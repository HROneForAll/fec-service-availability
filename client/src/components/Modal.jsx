import React from 'react';
import style from '../scripts/style.css.js';
import CheckIn from './CheckIn.jsx';
import CheckOut from './CheckOut.jsx';
import Calendar from './Calendar.jsx';

class Modal extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      calendar: false,
      start: null,
      end: null,
      weeks: [], 
      month: null,
      year: null
    };
  }

  render () {
    const displayModal = this.props.show ? style.showModal : style.hideModal;
    return (
      <div style={displayModal}>
        <section style={style.modalMain}>
          <button onClick={this.props.hide}>x</button>
          <div>
            <div>
              <label>
                <span>Dates</span>
              </label>
            </div>
            <div>
              <CheckIn click={() => this.showCalendar()}/>
              <div>-></div>
              <CheckOut click={() => this.showCalendar()}/>
              {this.state.calendar 
                ? <Calendar weeks={this.state.weeks}/> 
                : null}
            </div>
          </div>
          <button onClick={() => console.log(this.state)}>Test</button>
        </section>
      </div>
    );
  }

  showCalendar() {
    this.setState({calendar: true});
  }

  findFirstOfMonth(callback) {
    const today = new Date;
    let day = today.getDay();
    let date = today.getDate();
    while ((date - 1) % 7 !== 0 || date !== 1) {
      date--;
      day--;
      day < 0 ? day = 6 : day;
    }
    this.setState({ start: day });
    callback ? callback(day) : null;
  }

  findEndOfMonth(weeks) {
    const len = weeks.length - 1;
    for (let i = 0; i < 7; i++) {
      if (weeks[len][i] === null) {
        return i - 1;
      }
    }
  }

  setYearAndMonth() {
    const today = new Date;
    const year = today.getFullYear();
    const month = today.getMonth();
    this.setState({year, month});
  }

  buildWeeks(days, start, end, year, month) {
    const now = new Date; 
    year = year ? year : now.getFullYear();
    month = month ? month : now.getMonth();
    days = days ? days : Object.keys(this.props.dates[year][month]);
    if (start) {
      this.buildFromStart(days, start);
    } 
  }

  buildFromStart(days, start) {
    let weeks = [];
    let week = [];
    let count = 0;
    while (count < days.length + start) {
      if (count < start) {
        week.push(null);
      } else {
        week.push(days[count - start]);
      }
      if (week.length === 7) {
        weeks.push(week);
        week = [];
      }
      if (count === days.length + start - 1 && week.length < 7 && week.length) {
        while (week.length < 7) {
          week.push(null);
        }
        weeks.push(week);
      }
      count++;
    }
    const newEnd = this.findEndOfMonth(weeks);
    this.setState({ weeks, end: newEnd });
  }

  buildFromEnd(days, end) {
    let weeks = [];
    let week = [];
    let count = days.length + 6 - end;
    while (count > 0) {
      if (count > days.length) {
        week.unshift(null);
      } else {
        week.unshift(days[count - 1]);
      }
      if (week.length === 7) {
        weeks.unshift(week);
        week = [];
      }
      if (count === 1 && week.length && week.length < 7) {
        while (week.length < 7) {
          week.unshift(null);
        }
        weeks.unshift(week);
      }
      count--;
    }
    this.setState({ weeks });
  }

  componentDidMount() {
    this.findFirstOfMonth((start) => this.buildWeeks(null, start));
    this.setYearAndMonth();
  }

}

export default Modal;