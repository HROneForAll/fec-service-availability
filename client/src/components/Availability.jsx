import React from 'react';
import Navbar from './Navbar.jsx';
import axios from 'axios';
import Modal from './Modal.jsx';
import style from '../scripts/style.css.js';

class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      data: {},
      modal: false,
      listingId: 69
    };
  }

  render() {
    if (Object.keys(this.state.data).length) {
      const date = new Date;
      const id = this.state.listingId;
      const type = this.state.data.type;
      const location = this.state.data.location;
      const reviews = this.state.data.reviewNum;
      const stars = this.state.data.reviewSummary;
      const dates = this.state.data;
      const price = this.state.data[date.getFullYear()][date.getMonth()][date.getDate()]['price'];
      
      return (
        <div>
          <Navbar 
            type={type} 
            location={location} 
            reviews={reviews} 
            stars={stars}
            price={price} click={() => this.showModal()}
          />
          <Modal 
            show={this.state.modal} 
            hide={() => this.closeModal()}
            dates={dates} 
            reviews={reviews} 
            stars={stars} 
            price={price}
            id={id}
          />
        </div>
      );
    } else {
      return (
        <div>
          <h1>Loading...</h1>
        </div>
      );
    }
  }

  componentDidMount() {
    document.body.style.margin = '0';
    axios.get(`/availabilities/${this.state.listingId}`)
      .then(result => {
        const data = result.data[0].availability;
        this.setState({data});
      })
      .catch(err => console.log('Err in getting data', err));
  }

  showModal() {
    this.setState({modal: true});
  }

  closeModal() {
    this.setState({modal: false});
  }

}

export default App;