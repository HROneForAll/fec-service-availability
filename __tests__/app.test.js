import React from 'react';
import { shallow, mount, render } from 'enzyme';
import toJson from 'enzyme-to-json';
import renderer from 'react-test-renderer';
import Availability from '../client/src/components/Availability.jsx';
import Navbar from '../client/src/components/Navbar.jsx';
import Modal from '../client/src/components/Modal.jsx';
import Calendar from '../client/src/components/Calendar.jsx';

describe('Availability Render', () => {
  describe('Availability', () => {
    test('Renders the top-level component', () => {
      const wrapper =  shallow(<Availability />);
      expect(wrapper).toMatchSnapshot();
    });
  });

  const data = {
    type: 'Apartment',
    location: 'Houston',
    reviews: 199,
    stars: '4.7',
    price: 199,
    dates: {2019: {
      0: {1: {vacancy: true, price: 177}},
      1: {1: {vacancy: true, price: 177}
    }}},
    id: 69,
  }
  describe('Navbar', () => {
    test('Renders with data', () => {
      const wrapper = shallow(
        <Navbar 
          type={data.type} 
          location={data.location} 
          reviews={data.reviews} 
          stars={data.stars}
          price={data.price}
        />
      );
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Modal', () => {
    test('Renders with data', () => {
      const wrapper = shallow(
        <Modal
          dates={data.dates} 
          reviews={data.reviews} 
          stars={data.stars} 
          price={data.price}
          id={data.id}
        />
      );
      expect(wrapper).toMatchSnapshot();
    });

    test('Modal calendar should be hidden on load up', () => {
      const wrapper = shallow(<Modal />);
      expect(wrapper.state('calendar')).toEqual(false);
    });
  });

  const calendarData = {
    checkin: null,
    limit: null,
    checkout: null,
    dates: data.dates,
  }
  describe('Calendar', () => {
    test('Renders with data', () => {
      const wrapper = shallow(
        <Calendar 
          dates={calendarData.dates}
          checkin={calendarData.checkin}
          checkout={calendarData.checkout}
          limit={calendarData.limit}
        /> 
      );
      expect(wrapper).toMatchSnapshot();
    });

    test('Renders initial state with null/empty array and object', () => {
      const fn = jest.fn;
      const component = mount(
        <Calendar 
          dates={data.dates}
          checkin={data.in}
          checkout={data.out}
          limit={data.limit}
        /> 
      );
      component.find('.next-month-button').simulate('click');

    });
  });


});