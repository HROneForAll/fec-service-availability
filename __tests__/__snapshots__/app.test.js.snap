// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Render render() renders the component 1`] = `
<div>
  <h1>
    Loading...
  </h1>
</div>
`;
